pymc4/variational/approximations.py

```diff
@@ -105,64 +103,67 @@ class MeanField(Approximation):
     Inference. arXiv preprint arXiv:1603.00788.
     """
 
-    def _build_loc(self, shape, dtype, name):
-        loc = tf.Variable(tf.random.normal(shape, seed=self._seed), name=f"{name}/mu", dtype=dtype)
-        return loc
-
-    def _build_cov_matrix(self, shape, dtype, name):
-        # As per `tfp.vi.fit_surrogate_posterior` docs, use `TransformedVariable` or `DeferredTensor`
-        # to ensure all ops invoke gradients while applying transformation.
-        scale = tfp.util.TransformedVariable(
-            tf.fill(shape, value=tf.constant(1, dtype=dtype)),
-            tfb.Softplus(),  # For positive values of scale
-            name=f"{name}/sigma",
-        )
-        return scale
-
     def _build_posterior(self):
-        def apply_normal(dist_name):
-            unobserved_value = self.state.all_unobserved_values[dist_name]
-            shape = unobserved_value.shape
-            dtype = unobserved_value.dtype
-            return tfd.Normal(
-                self._build_loc(shape, dtype, dist_name),
-                self._build_cov_matrix(shape, dtype, dist_name),
-            )
-
-        # Should we use `tf.nest.map_structure` or `pm.utils.map_structure`?
-        variational_params = tf.nest.map_structure(apply_normal, self.unobserved_keys)
-        return tfd.JointDistributionSequential(variational_params)
+        flattened_shape = self.order.size
+        dtype = dtype_util.common_dtype(
+            self.state.all_unobserved_values.values(), dtype_hint=tf.float64
+        )
+        loc = tf.Variable(tf.random.normal([flattened_shape], dtype=dtype), name="mu")
+        cov_param = tfp.util.TransformedVariable(
+            tf.ones(flattened_shape, dtype=dtype), tfb.Softplus(), name="sigma"
+        )
+        advi_approx = tfd.MultivariateNormalDiag(loc=loc, scale_diag=cov_param)
+        return advi_approx
 
 
 class FullRank(Approximation):
-    """Full Rank Automatic Differential Variational Inference(Full Rank ADVI)."""
+    """
+    Full Rank ADVI.
 
-    def _build_loc(self):
-        raise NotImplementedError
+    This class implements Full Rank Automatic Differentiation Variational Inference. It posits Multivariate 
+    Gaussian family to fit posterior. And estimates a full covariance matrix. As a result, it comes with
+    higher computation costs.
 
-    def _build_cov_matrix(self):
-        raise NotImplementedError
+    References
+    ----------
+    -   Kucukelbir, A., Tran, D., Ranganath, R., Gelman, A.,
+    and Blei, D. M. (2016). Automatic Differentiation Variational
+    Inference. arXiv preprint arXiv:1603.00788.
+    """
 
     def _build_posterior(self):
-        raise NotImplementedError
+        flattened_shape = self.order.size
+        dtype = dtype_util.common_dtype(
+            self.state.all_unobserved_values.values(), dtype_hint=tf.float64
+        )
```

Comment:
Figuring out reasons of Cholesky Decomposition failure, this [comment](https://github.com/ICL-SML/Doubly-Stochastic-DGP/issues/9#issuecomment-371167725) has been very helpful. Here are my few observations experimenting with VI and Gaussian Processes - 
1. High learning rates (order of 1e-2) leads to cholesky decomposition errors.
2. Both Mean Field and Full Rank ADVI also lead to decomposition error with `tf.float32` even if we add a small WhiteNoise to the diagonal. But testing ADVI with `tf.float64` leads to greater numerical stability of covariance matrix. So, I changed the `dtype_hint` to use `tf.float64`.